src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,241 @@
+                    .response(pkc).build());
+            if (result.isSuccess() && result.isSignatureCounterValid()) {
+                final WebAuthnUserAccount account = webAuthnUserAccountRepository
+                        .findByUserHandle(result.getUserHandle().getBase64());
+                Set<WebAuthnCredential> credentials = account.getCredentials();
+                Optional<WebAuthnCredential> toUpdate = Optional.empty();
+                ByteArray resultCredentialId = result.getCredentialId();
+                for (WebAuthnCredential c : credentials) {
+                    ByteArray cCredentialId = c.getCredentialId();
+                    if (cCredentialId.equals(resultCredentialId)) {
+                        toUpdate = Optional.of(c);
+                    }
+                }
+                if (toUpdate.isEmpty()) {
+                    return Optional.empty();
+                }
+                WebAuthnCredential credential = toUpdate.get();
+                credentials.remove(credential);
+                credential.setSignatureCount(result.getSignatureCount());
+                credential.setLastUsedOn(new Date());
+                webAuthnCredentialsRepository.save(credential);
+                return Optional.of(account.getUsername());
+            }
+        } catch (Exception e) {
+        }
+        return Optional.empty();
+    }
+
+    UserIdentity getUserIdentityOrGenerate(String username, String realm, Optional<String> displayNameOpt,
+            Optional<Subject> optSub) {
+        String displayName = displayNameOpt.orElse("");
+        Optional<UserIdentity> option = getUserIdentity(username, displayName);
+        if (option.isPresent()) {
+            return option.get();
+        } else {
+            byte[] userHandle = new byte[64];
+            SecureRandom random = new SecureRandom();
+            random.nextBytes(userHandle);
+            final ByteArray userHandleBA = new ByteArray(userHandle);
+            final UserIdentity newUserIdentity = UserIdentity.builder()
+                    .name(username).displayName(displayName)
+                    .id(userHandleBA).build();
+            final WebAuthnUserAccount account = new WebAuthnUserAccount();
+            account.setUsername(username);
+            account.setCredentials(new HashSet<>());
+            account.setRealm(realm);
+            String subject;
+            if (optSub.isPresent()) {
+                subject = optSub.get().getSubjectId();
+            } else {
+                subject = subjectService.generateUuid(SystemKeys.RESOURCE_USER);
+            }
+            account.setSubject(subject);
+            account.setUserHandle(userHandleBA);
+            account.setProvider(provider);
+            webAuthnUserAccountRepository.save(account);
+            return newUserIdentity;
+        }
+    }
+
+    Optional<UserIdentity> getUserIdentity(String username, String displayName) {
```

Comment:
as before, no optional as return objects when empty is not acceptable. Also, our approach is: "get" methods for single entities throw exceptions when not found, while "find" or "fetch" return a null.

Reply:
Addressed in 141381a06e5fec8367e654b34341a7556a451790

Reply:
optional as return is still here.

Reply:
Fixed in 074890d7872c6fe888315cfa1d0c64a3c12b2e3d